lib/galaxy/datatypes/tabular.py

```diff
@@ -1380,3 +1381,75 @@ def set_meta(self, dataset, overwrite=True, skip=None, max_data_lines=7, **kwd):
             dataset.metadata.column_types = cleaned_column_types
             dataset.metadata.columns = number_of_columns
             dataset.metadata.delimiter = '\t'
+
+
+@dataproviders.decorators.has_dataproviders
+class DataFrame(TabularData):
+    """CSV or TSV hanled using pandas DataFrame.
+    """
+    file_ext = 'dataframe'
+
+    def set_meta(self, dataset, overwrite=True, skip=None, max_data_lines=10):
+        if not dataset.has_data():
+            log.warning("The dataset has empty data!")
+            return
+
+        sep = '\t'
+        try:
+            if CSV().sniff(dataset.file_name):
+                sep = ','
+        except Exception:
+            pass
+
+        dataset.metadata.delimiter = sep
+
+        with compression_utils.get_fileobj(dataset.file_name) as dataset_fh:
+            df = pd.read_csv(dataset_fh, nrows=0, sep=sep)
+            dataset.metadata.columns = df.shape[1]
+            # dataset.metadata.column_names = list(df.columns)
+            dataset_fh.seek(0)
+            dataset.metadata.data_lines = sum(1 for line in dataset_fh) - 1
+
+    def get_shape(self, dataset):
+        return (dataset.metadata.data_lines, dataset.metadata.columns)
+
+    def set_peek(self, dataset, peek_max_lines=5, peek_max_columns=10):
+        if not dataset.dataset.purged:
+            df = pd.read_csv(
+                dataset.file_name, sep=dataset.metadata.delimiter,
+                usecols=range(min(peek_max_columns, dataset.metadata.columns)),
+                nrows=peek_max_lines)
+            dataset.peek = df.to_json()
```

Comment:
This has the same problem that the `display_data` method below has, the json here can be arbitrarily large and take up a large amount of space on the peek column of the history_dataset_association table. The StringIO workaround here might be a good one though, and i like the idea of storing the peek in JSON, although I am not sure that JSON is the most efficient way to store tabular data, maybe csv would be better here.

Reply:
I can try the `to_string` method here.

Reply:
The `to_string` seems worse than `to_json`. :(

Reply:
That's why I suggested csv (using `to_csv()`)